src/pso.c

```diff
@@ -0,0 +1,204 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <string.h>
+
+#include <time.h>
+#include <float.h>
+
+#include "pso.h"
+#include "utils.h"
+
+
+#define EPS 0.001
+#define COG 2
+#define SOCIAL 2
+#define INERTIA 0.5
+#ifndef M_PI
+#define M_PI (3.14159265358979323846)
+#endif
+
+#define min(a, b) (((a) < (b)) ? (a) : (b))
+#define max(a, b) (((a) > (b)) ? (a) : (b))
+
+  //returns swarm_size * dim size array
+double* rand_init(size_t swarm_size,
+                  size_t dim,
+                  const double* const min_positions,
+                  const double* const max_positions) {
+  double* init = (double*)malloc(swarm_size*dim*sizeof(double));
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      init[idx] = random_min_max(min_positions[d],max_positions[d]);
+    }
+  }
+  return init;
+}
+
+  //returns swarm_size size array
+void eval_fitness(double (*func)(const double* const, size_t),
+                  size_t swarm_size, size_t dim,
+                  const double* const x, double* fitness) {
+  for (size_t pop=0; pop<swarm_size; pop++){
+    fitness[pop] = func(x+(pop*dim),dim);
+  }
+  return ;
+}
+
+  //returns swarm_size x dim size array
+  // velocity at every posize_t of every particle
+double* gen_init_velocity(const double* const x,
+                          size_t swarm_size, size_t dim,
+                          const double* const min_positions,
+                          const double* const max_positions) {
+  double* u = (double *)malloc(swarm_size*dim*sizeof(double));
+  u = rand_init(swarm_size,dim,min_positions,max_positions);
+  double* v = (double *)malloc(swarm_size*dim*sizeof(double));
+  for (size_t particle=0; particle < swarm_size; particle++){
+    for (size_t d=0; d<dim; d++){
+      size_t idx = (particle*dim) + d;
+      v[idx] = 0.25*(u[idx] - x[idx]);

```

Comment:
what is this hardcoded factor, should be defined outside the code at top of the file with descriptive name

Reply:
This will absolutely not change, it's pretty much in the algorithm. should I still define it outside?